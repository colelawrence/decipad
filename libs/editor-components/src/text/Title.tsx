import {
  ELEMENT_H1,
  PlateComponent,
  useTEditorState,
} from '@decipad/editor-types';
import { useIsEditorReadOnly } from '@decipad/react-contexts';
import { molecules } from '@decipad/ui';
import {
  findNodePath,
  focusEditor,
  isElementEmpty,
  select,
} from '@udecode/plate';
import { useEffect, useState } from 'react';

// TODO Title should probably not be a part of the editor in the first place

export const Title: PlateComponent = ({ attributes, children, element }) => {
  if (!element || element.type !== ELEMENT_H1) {
    throw new Error('Title is not a leaf');
  }

  const editor = useTEditorState();
  const [shouldAutofocus, setShouldAutofocus] = useState(true);
  const readOnly = useIsEditorReadOnly();

  useEffect(() => {
    if (shouldAutofocus) {
      const path = findNodePath(editor, element.children[0]);
      if (path) {
        select(editor, {
          path,
          offset: 0,
        });
        focusEditor(editor);
        setShouldAutofocus(false);
      }
    }
  }, [editor, element, shouldAutofocus]);

  return (
    <div {...attributes} contentEditable={!readOnly}>
      <molecules.EditorTitle
        Heading="h1"
        placeholder={
          isElementEmpty(editor, element) ? 'My notebook title' : undefined
        }
      >
        {children}
      </molecules.EditorTitle>
    </div>
  );
};
