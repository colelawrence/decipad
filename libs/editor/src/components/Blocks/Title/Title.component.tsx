import styled from '@emotion/styled';
import { SlatePluginComponent } from '@udecode/slate-plugins-core';
import React from 'react';

const Element = styled('h1')({
  fontSize: '2rem',
  margin: 0,
  paddingBottom: '2rem',
  marginBottom: '16px',
  borderBottom: '1px solid #f0f0f2',
  fontWeight: 'bold',
  color: '#121214',
  '& ::selection': {
    backgroundColor: 'rgba(196, 202, 251, 0.5)',
  },
});

export const Title: SlatePluginComponent = ({ attributes, children }) => {
  return <Element {...attributes}>{children}</Element>;
};
