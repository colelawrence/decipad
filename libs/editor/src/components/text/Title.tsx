import { atoms } from '@decipad/ui';
import { useEditorState } from '@udecode/plate';
import { useEffect, useState } from 'react';
import { Editor, Transforms } from 'slate';
import { ReactEditor } from 'slate-react';
import { PlateComponent } from '../../types';
import { ELEMENT_H1 } from '../../elements';

// TODO Title should probably not be a part of the editor in the first place

export const Title: PlateComponent = ({ attributes, children, element }) => {
  if (!element || element.type !== ELEMENT_H1) {
    throw new Error('Title is not a leaf');
  }

  const editor = useEditorState();
  const [shouldAutofocus, setShouldAutofocus] = useState(true);

  useEffect(() => {
    if (shouldAutofocus) {
      Transforms.select(editor, {
        path: ReactEditor.findPath(editor, element.children[0]),
        offset: 0,
      });
      ReactEditor.focus(editor);
      setShouldAutofocus(false);
    }
  }, [editor, element, shouldAutofocus]);

  return (
    <div {...attributes}>
      <atoms.Display
        Heading="h1"
        placeholder={
          Editor.isEmpty(editor, element) ? 'My notebook title' : undefined
        }
      >
        {children}
      </atoms.Display>
      <div contentEditable={false} css={{ paddingBottom: '16px' }}>
        <atoms.Divider />
      </div>
    </div>
  );
};
