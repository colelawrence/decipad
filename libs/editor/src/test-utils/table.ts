import {
  ELEMENT_TABLE,
  ELEMENT_TH,
  ELEMENT_TR,
  ELEMENT_TD,
} from '@udecode/plate';
import {
  InteractiveTable,
  TableCellType,
  Td,
  Th,
} from '../plugins/InteractiveTable/table';
import {
  ELEMENT_TABLE_CAPTION,
  ELEMENT_THEAD,
  ELEMENT_HEAD_TR,
  ELEMENT_TBODY,
} from '../utils/elementTypes';

/**
 * Creates a row with 1 named column and {@param numberOfRows} rows of non-empty cells.
 */
export const createTable = (numberOfRows: number): InteractiveTable => ({
  type: ELEMENT_TABLE,
  children: [
    {
      type: ELEMENT_TABLE_CAPTION,
      children: [{ text: 'Table' }],
    },
    {
      type: ELEMENT_THEAD,
      children: [
        {
          type: ELEMENT_HEAD_TR,
          children: [
            {
              type: ELEMENT_TH,
              attributes: { cellType: 'string' },
              children: [{ text: 'Col 1' }],
            },
          ],
        },
      ],
    },
    {
      type: ELEMENT_TBODY,
      children: Array.from({ length: numberOfRows }, (_, i) => ({
        type: ELEMENT_TR,
        children: [
          {
            type: ELEMENT_TD,
            attributes: { cellType: 'string' },
            children: [{ text: `Col 1 Row ${i + 1}` }],
          },
        ],
      })),
    },
  ],
});

export const createTableWithTypeAndData = (
  cellType: TableCellType,
  data: string[]
): InteractiveTable => ({
  type: ELEMENT_TABLE,
  children: [
    {
      type: ELEMENT_TABLE_CAPTION,
      children: [{ text: 'Table' }],
    },
    {
      type: ELEMENT_THEAD,
      children: [
        {
          type: ELEMENT_HEAD_TR,
          children: [
            {
              type: ELEMENT_TH,
              attributes: { cellType },
              children: [{ text: 'Col 1' }],
            },
          ],
        },
      ],
    },
    {
      type: ELEMENT_TBODY,
      children: data.map((text) => ({
        type: ELEMENT_TR,
        children: [
          {
            type: ELEMENT_TD,
            attributes: { cellType },
            children: [{ text }],
          },
        ],
      })),
    },
  ],
});

export const printTable = ({
  children: [_, thead, tbody],
}: InteractiveTable): string[] => {
  const tdsToStrings = (tds: (Th | Td)[]) =>
    tds.map((td) => td.children[0].text).join(' | ');

  return [
    tdsToStrings(thead.children[0].children),
    ...tbody.children.map((tr) => tdsToStrings(tr.children)),
  ];
};
