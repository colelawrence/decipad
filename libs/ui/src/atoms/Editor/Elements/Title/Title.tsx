import { css } from '@emotion/react';
import { PlatePluginComponent } from '@udecode/plate';
import { cssVar, display } from '../../../../primitives';

const styles = css(display, {
  borderBottom: '1px solid',
  borderColor: cssVar('highlightColor'),
  paddingBottom: '24px',
  marginBottom: '16px',
});

export const TitleElement: PlatePluginComponent = ({
  attributes,
  children,
  nodeProps,
}) => {
  // TODO type root better and make optional
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const { root } = nodeProps!.styles;
  return (
    <h1 css={[styles, root.css]} {...attributes} {...nodeProps}>
      {children}
    </h1>
  );
};
