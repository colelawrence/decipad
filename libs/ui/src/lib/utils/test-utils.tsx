import { ChakraProvider } from '@chakra-ui/react';
import { render, RenderOptions } from '@testing-library/react';
import React from 'react';

import { Type } from '@decipad/language';
import { theme, ResultsContextProvider } from '../..';

const results = {
  test: {
    type: Type.build({
      type: 'number',
      unit: {
        unit: 'apples',
      },
    } as any),
    value: [20],
    errors: [],
  },
  noResult: null,
};

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <ChakraProvider resetCSS theme={theme}>
      <ResultsContextProvider value={results}>
        {children}
      </ResultsContextProvider>
    </ChakraProvider>
  );
};

const customRender = (
  ui: JSX.Element,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
