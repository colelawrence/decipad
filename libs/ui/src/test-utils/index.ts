export * from './menu';
export * from './language';
