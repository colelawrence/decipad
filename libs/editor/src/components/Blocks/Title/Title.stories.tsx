import React from 'react';
import { Title } from './Title.component';

export default {
  title: 'Editor/Blocks/Title',
  component: Title,
};

export const Default = (args: { text: string }) => (
  <Title
    leaf={{ attributes: {}, text: '' }}
    element={{ children: [{ text: '' }] }}
    text={{ text: '' }}
    attributes={{
      'data-slate-node': 'element',
      ref: null,
      'data-slate-leaf': true,
    }}
  >
    {args.text}
  </Title>
);

Default.args = {
  text: 'Title',
};

Default.storyName = 'Title';
