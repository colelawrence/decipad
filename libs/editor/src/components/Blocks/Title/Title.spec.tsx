import React from 'react';
import { render } from 'test-utils';
import { Default } from './Title.stories';

describe('Title Block', () => {
  it('matches snapshot', () => {
    const { container } = render(<Default text="Title" />);
    expect(container).toMatchSnapshot();
  });
});
