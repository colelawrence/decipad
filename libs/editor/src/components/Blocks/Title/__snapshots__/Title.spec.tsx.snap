// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title Block matches snapshot 1`] = `
.emotion-0 {
  font-size: 2rem;
  margin: 0;
  padding-bottom: 2rem;
  margin-bottom: 16px;
  border-bottom: 1px solid #f0f0f2;
  font-weight: bold;
  color: #121214;
}

.emotion-0 ::selection {
  background-color: rgba(196, 202, 251, 0.5);
}

<div>
  <h1
    class="emotion-0"
    data-slate-leaf="true"
    data-slate-node="element"
  >
    Title
  </h1>
</div>
`;
